sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb

```diff
@@ -0,0 +1,574 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Set up the environment\n",
+    "\n",
+    "We will fetch the `IAM` role this notebook is running in and pass the same role to the TensorFlow estimator which SageMaker will assume to fetch data and perform training."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import sagemaker\n",
+    "import os\n",
+    "from sagemaker.utils import sagemaker_timestamp\n",
+    "from sagemaker.tensorflow import TensorFlow\n",
+    "from sagemaker import get_execution_role\n",
+    "\n",
+    "sagemaker_session = sagemaker.Session()\n",
+    "\n",
+    "default_s3_bucket = sagemaker_session.default_bucket()\n",
+    "sagemaker_iam_role = get_execution_role()\n",
+    "\n",
+    "train_script = \"mnist_hvd.py\"\n",
+    "instance_count = 2"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Prepare Data for training\n",
+    "\n",
+    "In this, we will download the mnist dataset to the local `/tmp/data/` directory which is then uploaded to the s3 followed by cleaning up `/tmp/data/`. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "import shutil\n",
+    "\n",
+    "import numpy as np\n",
+    "\n",
+    "import keras\n",
+    "from keras.datasets import mnist\n",
+    "(x_train, y_train), (x_test, y_test) = mnist.load_data()\n",
+    "\n",
+    "s3_train_path = \"s3://{}/mnist/train.npz\".format(default_s3_bucket)\n",
+    "s3_test_path = \"s3://{}/mnist/test.npz\".format(default_s3_bucket)\n",
+    "\n",
+    "# Create local directory\n",
+    "! mkdir -p /tmp/data/mnist_train\n",
+    "! mkdir -p /tmp/data/mnist_test\n",
+    "\n",
+    "# Save data locally\n",
+    "np.savez('/tmp/data/mnist_train/train.npz', data=x_train, labels=y_train)\n",
+    "np.savez('/tmp/data/mnist_test/test.npz', data=x_test, labels=y_test)\n",
+    "\n",
+    "# Upload the dataset to s3\n",
+    "! aws s3 cp /tmp/data/mnist_train/train.npz $s3_train_path\n",
+    "! aws s3 cp /tmp/data/mnist_test/test.npz $s3_test_path\n",
+    "\n",
+    "print('training data at ', s3_train_path)\n",
+    "print('test data at ', s3_test_path)\n",
+    "! rm -rf /tmp/data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Construct a script for horovod distributed training\n",
+    "\n",
+    "In this example we are using the [Keras MNIST horovod example](https://github.com/uber/horovod/blob/master/examples/keras_mnist.py) from the horovod github repository.\n",
+    "\n",
+    "In order to run this script we have to make following modifications:"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 1. Accept `--model_dir` command-line argument\n",
```

Comment:
Do you mind detailing here that model_dir point to /opt/ml/model which is the special folder that SageMaker saves in the end of training and that works for serving as well. Please mentioned that it allows the training job to integrates with other SageMaker services and avoid the model to be lost after training given that the training job is ephemeral.

Please add some links as well.

Reply:
Done